src/main/scala/viper/silver/plugin/standard/inline/InlineRewrite.scala

```diff
@@ -0,0 +1,79 @@
+package viper.silver.plugin.standard.inline
+
+import scala.collection.mutable
+import viper.silver.ast._
+import viper.silver.ast.utility.ViperStrategy
+import viper.silver.ast.utility.rewriter.Traverse
+
+trait InlineRewrite {
+
+  /**
+    * Rewrites a method by expanding the predicates in its pre- and postconditions,
+    * then by removing the corresponding unfold and fold statements in the body.
+    * Note that we only remove unfolds for predicates from the preconditions,
+    * and folds for predicates from the postconditions.
+    * (I think that's how it works.)
+    */
+  def rewriteMethod(method: Method, program: Program): Method = {
+    val (expandedPres, expandablePrePredsIds) = method.pres
+      .map { expandPredicates(_, method, program) }
+      .unzip
+    val (expandedPosts, expandablePostPredsIds) = method.posts
+      .map { expandPredicates(_, method, program) }
+      .unzip
+    val expandablePrePredIds = expandablePrePredsIds.flatten.toSet
+    val expandablePostPredIds = expandablePostPredsIds.flatten.toSet
+    val rewrittenBody = method.body
+      .map { removeFoldUnfolds(_, expandablePrePredIds, expandablePostPredIds) }
+    method.copy(body = rewrittenBody,
+      pres = expandedPres,
+      posts = expandedPosts,
+    )(method.pos, method.info, method.errT)
+  }
+
+  /**
+    * Expands all predicates to their bodies in given expression.
+    *
+    * @param expr The expression whose predicates will be expanded
+    * @param method The method containing the expression, used to determine locally-scoped variables
+    * @param program The program containing the expression, used to expand predicates
+    * @return The expression with expanded predicates and a set of their names
+    */
+  private[this] def expandPredicates(expr: Exp, method: Method, program: Program): (Exp, Set[String]) = {
+    // Forgive me deities of functional programming for I sin
+    val expandablePredicates = mutable.Set[String]()
+    val expandedExpr = ViperStrategy.Context[Set[String]]({
+      case exp@(PredicateAccessPredicate(pred, _), ctxt) =>
+        val body = pred.predicateBody(program, ctxt.c)
+        val isInUnfolding = ctxt.parentOption.fold(false)(_.isInstanceOf[Unfolding])
```

Comment:
You may not like it, but this is peak monad
```Scala
val isInUnfolding = ctxt.parentOption.exists {
    case _: Unfolding => true
}
```

Reply:
Ah, but:
```scala
val isInUnfolding = ctxt.parentOption.exists({case Unfolding(_, _) => true})
```

Reply:
😩

Reply:
Oops, this is broken bc it's missing `case _ => false` :facepalm:

Reply:
Is it a compiler error? I think `exists` is false for everything else that doesn't satisfy the predicate...

Reply:
nvm, the function passed to `exists` can't be partial 😩

Reply:
`sbt` didn't warn me :angry:

Reply:
I strongly suspect it's because they didn't use a `sealed trait` to define the top-level types; this actually tells the compiler to yell at the dev for not handling all cases of what the thing could be.

Smh my head Viper developers